src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs

```diff
@@ -288,56 +324,86 @@ private void UpdateStreamTimeouts(DateTimeOffset now)
                     Debug.Assert(streamDirectionFeature != null);
                     Debug.Assert(streamIdFeature != null);
 
+                    var context = CreateHttpStreamContext(streamContext);
+                    var pendingStream = new Http3PendingStream(context, streamIdFeature.StreamId);
+
+                    _streamLifetimeHandler.OnUnidentifiedStreamReceived(pendingStream);
+
+                    // unidirectional stream
                     if (!streamDirectionFeature.CanWrite)
                     {
-                        // Unidirectional stream
-                        var stream = new Http3ControlStream<TContext>(application, CreateHttpStreamContext(streamContext));
-                        _streamLifetimeHandler.OnStreamCreated(stream);
+                        var streamType = await pendingStream.ReadNextStreamHeaderAsync(context, streamIdFeature.StreamId, null);
```

Comment:
This is a problem. The loop to accept streams should do as little work as possible. Reading from the stream here blocks accepting new streams. For example, a stream that delays sending data for 20 seconds blocks the connection from accepting more streams for that time.

Reply:
Is there some other way of knowing the stream type? I was under the impression that this is only accessible via the quic integer that is the first 2 bytes on the stream

Reply:
Yes, it looks like the stream type is read from the stream content: https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-03.html#section-4

Reading stream content has to happen after it's dispatched to the threadpool via `ThreadPool.UnsafeQueueUserWorkItem`. Otherwise, it blocks accepting new streams.

`Http3ControlStream` already has logic that [reads a header type](https://github.com/dotnet/aspnetcore/blob/44b3ddd19216a31c9ace4b6f8b8fb74f0643be44/src/Servers/Kestrel/Core/src/Internal/Http3/Http3ControlStream.cs#L155). `Http3ControlStream` has different processing depending on the stream type (control vs encoder vs decoder).

`Http3Stream` is more hardcoded to reading HTTP/3 frames, but a frame header is pretty similar to a stream header (type + length compared to WebTransports type + session). The spec even says it [sends a HTTP/3 frame to identify the WebTransport stream](https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-03.html#section-4.2). `Http3Stream` could be made more like the control stream and have different content processing depending on the stream type. If it gets a standard headers frame then treat the stream as a normal HTTP/3 request. If it gets a WebTransport frame then treat the stream as WebTransport. We'd probably want to move a lot of the request processing logic in `Http3Stream` to another type. For example, there is a lot of code for processing response headers and writing a valid H3 response that's not relevant to WebTransport.

Reply:
Actually, another temporary solution could be to use the webtransport `AppContextSwitch` here and only use the pending stream concept when webtransport is enabled. Otherwise just default to the previous method of using `CanRead` and `CanWrite`. Would this be a good work around? It wouldn't be much code and the result will be that non webtransport traffic will be almost exactly the same as before my changes (just a couple extra ifs/conditions to existing ifs will need to be added).

Reply:
Yes, use IThreadPoolItem and UnsafeQueueUserWorkItem to dispatch pending streams off the main look. That should unblock the main issue here. 

Using the AppContextSwitch would be hard because of how much has had to be restructured like with the control streams.

Reply:
Turns out that the converse is true. The `AppContextSwitch` was actually way easier than the `UnsafeQueueUserWorkItem` as the `UnsafeQueueUserWorkItem` approach would create complexity (that would become technical debt) in synchronizing the pending stream and the connection. The `AppContextSwitch` avoids all that.